### PR TITLE
wp_trim_excerpt parses and renders blocks twice  also avoid unwanted block parsing

### PR DESCRIPTION
**Pull request by kt-12**
#### Performance benchmark with `trunk`

| Metric                       | trunk   | After PR |
|------------------------------|---------|----------|
| **Response Time (median)**   | 164.52  | 151.95   |
| **wp-before-template (median)** | 62.44  | 59.18    |
| **wp-template (median)**     | 98.5    | 89.83    |
| **wp-total (median)**        | 161.7   | 149.38   |

Trac ticket: [https://core.trac.wordpress.org/ticket/58682](https://core.trac.wordpress.org/ticket/58682)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/formatting.php [hunk 1]
```diff
@@ -3980,18 +3980,29 @@ function wp_trim_excerpt( $text = '', $post = null ) {
 		 * within the excerpt are stripped out. Modifying the tags here
 		 * is wasteful and can lead to bugs in the image counting logic.
 		 */
-		$filter_removed = remove_filter( 'the_content', 'wp_filter_content_tags' );
+		$filter_image_removed = remove_filter( 'the_content', 'wp_filter_content_tags' );
+
+		/**
```
Comment (mukeshpanchal27) [inline src/wp-includes/formatting.php]:
```suggestion
		/*
```

Accidental extra `*` here.

File: src/wp-includes/formatting.php [hunk 2]
```diff
@@ -3980,18 +3980,29 @@ function wp_trim_excerpt( $text = '', $post = null ) {
...
 		/** This filter is documented in wp-includes/post-template.php */
 		$text = apply_filters( 'the_content', $text );
 		$text = str_replace( ']]>', ']]&gt;', $text );
 
+		// Restore the original filter if removed.
+		if ( $filter_block_removed ) {
+			add_filter( 'the_content', 'do_blocks', 9 );
+		}
+
 		/**
```
Comment (mukeshpanchal27) [inline src/wp-includes/formatting.php]:
```suggestion
		/*
```

See above for more details

File: tests/phpunit/tests/formatting/wpTrimExcerpt.php [hunk 3]
```diff
@@ -148,4 +148,77 @@ public function test_wp_trim_excerpt_does_not_restore_wp_filter_content_tags_if_
 		// Assert that the filter callback was not restored after running 'the_content'.
 		$this->assertFalse( has_filter( 'the_content', 'wp_filter_content_tags' ) );
 	}
+
+	/**
+	 * Tests that `wp_trim_excerpt()` does process valid blocks.
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/formatting/wpTrimExcerpt.php:153]: Can we add a `@covers` annotation to each test?

**Review by spacedmonkey — APPROVED** [review]
This improve performance in my testing and the unit tests look good to me.

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @kt-12 for the PR. Left some initial feedback.

**Comment by mukeshpanchal27** [comment]
@kt-12, I'm curious to know which theme you use in your benchmark testing. Could you please share the data for both block and classic themes?

**Comment by spacedmonkey** [comment]
@mukeshpanchal27 Have you seen - https://github.com/WordPress/wordpress-develop/actions/runs/6159089777

**Review by felixarntz — APPROVED** [review]
Thanks @kt-12 for the PR, this looks excellent and IMO good to commit!

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/56560

**Pull request closed without merging**